ocrd/ocrd/workspace.py

```diff
@@ -461,64 +417,124 @@ def image_from_page(self, page, page_id,
-            if not (page_image.width == page_xywh['w'] and
-                    page_image.height == page_xywh['h']):
-                log.error('page "%s" image (%s; %dx%d) has not been transposed properly (%dx%d) during rotation',
+        # adjust the coord transformation to the steps applied on the image,
+        # and apply steps on the existing image in case it is missing there,
+        # but traverse all steps (crop/reflect/rotate) in a particular order:
+        # - existing image features take priority (in the order annotated),
+        # - next is cropping (if necessary but not already applied),
+        # - next is reflection (if necessary but not already applied),
+        # - next is rotation (if necessary but not already applied).
+        # This helps deal with arbitrary workflows (e.g. crop then deskew,
+        # or deskew then crop), regardless of where images are generated.
+        alternative_image_features = page_coords['features'].split(',')
+        for i, feature in enumerate(alternative_image_features +
+                                    (['cropped']
+                                     if (border and
+                                         not 'cropped' in page_coords['features'] and
+                                         not 'cropped' in feature_filter.split(','))
+                                     else []) +
+                                    (['rotated-%d' % orientation]
+                                     if (orientation and
+                                         not 'rotated-%d' % orientation in page_coords['features'] and
+                                         not 'rotated-%d' % orientation in feature_filter.split(','))
+                                     else []) +
+                                    (['deskewed']
+                                     if (skew and
+                                         not 'deskewed' in page_coords['features'] and
+                                         not 'deskewed' in feature_filter.split(','))
+                                     else []) +
+                                    # not a feature to be added, but merely as a fallback position
+                                    # to always enter loop at i == len(alternative_image_features)
+                                    ['_check']):
+            # image geometry vs feature consistency can only be checked
+            # after all features on the existing AlternativeImage have
+            # been adjusted for in the transform, and when there is a mismatch,
+            # additional steps applied here would only repeat the respective
+            # error message; so we only check once at the boundary between
+            # existing and new features
+            # FIXME we should check/enforce consistency when _adding_ AlternativeImage
+            if (i == len(alternative_image_features) and
+                not (page_xywh['w'] - 2 < page_image.width < page_xywh['w'] + 2 and
```

Comment:
A constant `PAGE_CROP_TOLERANCE` might make the intention clearer than `2`.

Reply:
Indeed.